invisible_cities/cities/components.py

```diff
@@ -594,12 +594,17 @@ def hits_and_kdst_from_files( paths : List[str]
             except (tb.exceptions.NoSuchNodeError, IndexError):
                 continue
 
-            check_lengths(event_info, hits_df.event.unique())
+            check_lengths(event_info, kdst_df.event.unique())
 
             for evtinfo in event_info:
                 event_number, timestamp = evtinfo.fetch_all_fields()
                 hits = hits_from_df(hits_df.loc[hits_df.event == event_number])
-                yield dict(hits = hits[event_number],
+                if len(hits):
+                    hits = hits[event_number]
+                else:
+                    warnings.warn(f"Event {event_number} does not contain hits", UserWarning)
+                    hits = HitCollection(event_number, timestamp, [])
+                yield dict(hits = hits,
```

Comment:
`if len(hits):` returns `True` for any number of rows in the `hits` dataframe, and `False` only for 0 rows, is that correct? This usage of `if len(hits):` interpreting the length of a dataframe like this feels a bit obtuse to me, but perhaps that is due to my lack of familiarity with boolean behaviour in python.

Reply:
Yes, this is idiomatic in Python. When interpreted as booleans, all numbers except 0 are `True` (even `np.nan`), and 0 is `False`.
This pattern is particularly common in sequences. `if len(something)` means "if it's not empty". You may also see `if not len(something)`, which means "if it's empty".
In this particular case, `hits` is a dictionary that maps event numbers to `HitCollection`s. The `len` of a dictionary corresponds to the number of entries and therefore the meaning above is preserved.

Reply:
Idiomatic Python would omit the `len`. Empty containers (be they sequences, mappings, whatever) are false in boolean contexts in Python and you are expected to take advantage of this ... except that numpy arrays throw a spanner in the works by being sequences in most respects, with the notable exception of their boolean interpretation. Consequently, in environments where numpy is used, it becomes less clear that the standard Python idioms are a good idea:

```python
import numpy as np

samples = (
                 ''  ,              'a'  ,
                 []  ,              [0]  ,
                 ()  ,              (0,) ,
                 {}  ,              {0:0},
              set()  ,              {0}  ,
    #   np.array([]) ,     np.array([0]) , # Oh dear, numpy is a git
    len(np.array([])), len(np.array([0])), # `len` provides a workaround
)

def bbb(x):
	print(x, bool(x))
    
for x in samples:
	bbb(x)
```

So, while the superfluous-for-all-containers-except-numpy-arrays `len`, is bad style in Python in general, it might be justified in projects that use numpy, for the sake of consistency and avoiding numpy-caused surprises.